text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,455 @@
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: '12345', region: 'us-east-1' },
+});
+pipeline.addCdkStage(stage, {
+  after: [
+    new cdkp.ShellScriptAction({
+      input: pipeline.build.additionalOutput('tests'),
+      commands: ['node ./integ-tests'],
+    }),
+  ],
+});
+```
+
+### Splitting definition and backend
+
+It is helpful if there is a clearly defined moment in time when the pipeline definition is ready to be rendered.
+
+There are various ways to organize this, each pretty much isomorphic to one another.
+
+OPTION 1
+
+Clear split of definition and instantiation. Very much like StepFunctions.
+
+```ts
+const pipelineDefinition = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    ...
+  }),
+});
+pipelineDefinition.addCdkStage(...);
+pipelineDefinition.addCdkStage(...);
+
+new cdkp.CodePipelineBackend(this, 'Pipeline', {
+  pipelineDefinition: pipelineDefinition,
+});
+```
+
+OPTION 2
+
+Automatic rendering moment.
+
+```ts
+const pipeline = new cdkp.Pipeline(this, 'Pipeline', {
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    ...
+  }),
+  backend: new cdkp.CodePipelineBackend({
+    ...backendOptions
+}),
+});
+pipeline.addCdkStage(...);
+pipeline.addCdkStage(...);
+
+pipeline.render();  // <-- can be automatically called if not called at `prepare()` time.
+```
+
+OPTION 3 (syntactic sugar over OPTIONS 1 or 2)
+
+Backend-specific syntactic sugar for people that definitely only need one backend.
+
+```ts
+const pipeline = new cdkp.CodePipeline(this, 'Pipeline', {
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    ...
+  }),
+  ...backendOptions
+});
+pipeline.addCdkStage(...);
+pipeline.addCdkStage(...);
+
+pipeline.render();  // <-- can be automatically called if not called at `prepare()` time.
+```
+
+Advantages of options 2, 3 are that `render()` can be implicit if necessary, and so cannot
+be accidentally forgotten.
+
+Recommendation: we implement options 1 and 3.
+
+### Waves
+
+Deploying multiple applications in parallel is done by adding waves:
+
+```ts
+const def = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({ ... }),
+  waves: [
+    {
+      waveName: '...',
+      stages: [
+        {
+          stageName?: '...',
+          stage: new MyAppStage(...),
+          after: [...],
+        }
+      ],
+    },
+  ],
+});
+def.addCdkStage(new MyAppStage(...));
+
+const wave = def.addWave('asdf', {
+  stages: [...],
+});
+wave.addCdkStage(new MyAppStage(...));
+```
+
+### Validations
+
+Waves and Stages can be decorated with `before` and `after` actions, intended for validations
+and checks.
+
+```ts
+const def = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({ ... }),
+});
+def.addCdkStage(new MyAppStage('Beta', ...));
+
+const wave = def.addWave('Prod', {
+  before: [
+    new cdkp.ManualApprovalAction('Promote Beta to Prod'),
```

Comment:
Just to get away from CodePipeline terminology, maybe we can rename "Actions" to "Tasks"? or "Steps"?

Reply:
Yes I actually meant to call it `Step`.